src/goose/brokers/redis/batch.clj

```diff
@@ -70,3 +58,72 @@
+                       d/redis-batch-lock-timeout-ms
+                       d/redis-batch-lock-wait-ms
+                       (let [{:keys [batch-state] :as batch} (get-batch-state conn id)
+                             status (batch/status-from-counts batch)
+                             {:keys [callback-job-id
+                                     callback-fn-sym
+                                     linger-in-hours
+                                     queue
+                                     ready-queue
+                                     retry-opts]} batch-state]
+                         (when (= status batch/status-complete)
+                           (when (and
+                                   (some? callback-fn-sym)
+                                   (not callback-job-id))
+                             (let [callback-args (-> (select-keys batch [:successful :dead])
+                                                     (list)
+                                                     (conj id))
+                                   callback-job (-> (job/new callback-fn-sym
+                                                             callback-args
+                                                             queue
+                                                             ready-queue
+                                                             retry-opts)
+                                                    (assoc :batch-id id))]
+                               (redis-cmds/enqueue-back conn ready-queue callback-job)
+                               (redis-cmds/hset conn (d/prefix-batch id) (name :callback-job-id) (:id callback-job))))
+                           (let [linger-in-hours (if linger-in-hours
+                                                   (Integer/parseInt linger-in-hours)
+                                                   d/redis-batch-linger-hours)]
+                             (set-batch-expiration conn id linger-in-hours))))))
+
+(defn wrap-state-update [next]
+  (fn [{:keys [redis-conn] :as opts}
+       {:keys [id batch-id] :as job}]
+    (if batch-id
+      (let [src (if (job/retried? job)
+                  (d/construct-batch-job-set batch-id d/retrying-job-set)
+                  (d/construct-batch-job-set batch-id d/enqueued-job-set))]
+        (try
+          (let [response (next opts job)
+                dst (d/construct-batch-job-set batch-id d/successful-job-set)]
+            (redis-cmds/move-between-sets redis-conn src dst id)
+            (post-batch-exec redis-conn batch-id)
+            response)
+          (catch Exception ex
+            (let [failed-job (goose.retry/set-failed-config job ex)
+                  dst (if (goose.retry/max-retries-reached? failed-job)
+                        (d/construct-batch-job-set batch-id d/dead-job-set)
+                        (d/construct-batch-job-set batch-id d/retrying-job-set))]
+              (redis-cmds/move-between-sets redis-conn src dst id)
```

Comment:
`(post-batch-exec redis-conn batch-id)` must be called here as well. I'm wondering if moving move-between-sets & post-batch-exec to a finally block or a common function makes sense.

If we're moving it to finally block, we need to check if `response` gets returned at end of this block or the result of finally block.